src/field.rs

```diff
@@ -15,12 +15,83 @@ use std::ops::Deref;
 use std::os::raw::c_void;
 use tibrv_sys::*;
 
+pub enum DecodedField<'m> {
+    String(&'m CStr),
+    Message(BorrowedMsg),
+    U8(u8),
+    U8Array(&'m [u8]),
+    I8(i8),
+    I8Array(&'m [i8]),
+    U16(u16),
+    U16Array(&'m [u16]),
+    I16(i16),
+    I16Array(&'m [i16]),
+    U32(u32),
+    U32Array(&'m [u32]),
+    I32(i32),
+    I32Array(&'m [i32]),
+    U64(u64),
+    U64Array(&'m [u64]),
+    I64(i64),
+    I64Array(&'m [i64]),
+    F32(f32),
+    F32Array(&'m [f32]),
+    F64(f64),
+    F64Array(&'m [f64]),
+    Bool(bool),
+    DateTime(NaiveDateTime),
+    Ipv4(Ipv4Addr),
+    IpPort(u16),
+    Opaque(&'m [u8]),
+}
+
+impl<'m> Decodable<'m> for DecodedField<'m> {
+    fn tibrv_try_decode(fld: &'m MsgField) -> Result<DecodedField<'m>, TibrvError> {
+        match fld.inner.type_ as u32 {
+            TIBRVMSG_STRING => fld.try_decode().map(DecodedField::String),
+            TIBRVMSG_MSG => fld.try_decode().map(DecodedField::Message),
+            TIBRVMSG_U8 => fld.try_decode().map(DecodedField::U8),
+            TIBRVMSG_U8ARRAY => fld.try_decode().map(DecodedField::U8Array),
+            TIBRVMSG_I8 => fld.try_decode().map(DecodedField::I8),
+            TIBRVMSG_I8ARRAY => fld.try_decode().map(DecodedField::I8Array),
+            TIBRVMSG_U16 => fld.try_decode().map(DecodedField::U16),
+            TIBRVMSG_U16ARRAY => fld.try_decode().map(DecodedField::U16Array),
+            TIBRVMSG_I16 => fld.try_decode().map(DecodedField::I16),
+            TIBRVMSG_I16ARRAY => fld.try_decode().map(DecodedField::I16Array),
+            TIBRVMSG_U32 => fld.try_decode().map(DecodedField::U32),
+            TIBRVMSG_U32ARRAY => fld.try_decode().map(DecodedField::U32Array),
+            TIBRVMSG_I32 => fld.try_decode().map(DecodedField::I32),
+            TIBRVMSG_I32ARRAY => fld.try_decode().map(DecodedField::I32Array),
+            TIBRVMSG_U64 => fld.try_decode().map(DecodedField::U64),
+            TIBRVMSG_U64ARRAY => fld.try_decode().map(DecodedField::U64Array),
+            TIBRVMSG_I64 => fld.try_decode().map(DecodedField::I64),
+            TIBRVMSG_I64ARRAY => fld.try_decode().map(DecodedField::I64Array),
+            TIBRVMSG_F32 => fld.try_decode().map(DecodedField::F32),
+            TIBRVMSG_F32ARRAY => fld.try_decode().map(DecodedField::F32Array),
+            TIBRVMSG_F64 => fld.try_decode().map(DecodedField::F64),
+            TIBRVMSG_F64ARRAY => fld.try_decode().map(DecodedField::F64Array),
+            TIBRVMSG_BOOL => fld.try_decode().map(DecodedField::Bool),
+            TIBRVMSG_DATETIME => fld.try_decode().map(DecodedField::DateTime),
+            TIBRVMSG_IPADDR32 => fld.try_decode().map(DecodedField::Ipv4),
+            TIBRVMSG_IPPORT16 => tibrv_try_decode_port(fld).map(DecodedField::IpPort),
+            TIBRVMSG_OPAQUE => unsafe { tibrv_try_decode_opaque::<u8>(fld).map(DecodedField::Opaque) },
+            _ => Err(ErrorKind::FieldTypeError.into()), // FIXME: should be more specific error type
```

Comment:
I suggest `ErrorKind::UnknownFieldType`, it's probably never going to happen so you're right we're better to throw an error than include an `Unknown` variant or such.

Maybe we want to rename `FieldTypeError` to be more specific then?